### PR TITLE
Readonly members warnings and semantic errors

### PR DESCRIPTION
**Pull request by RikkiGibson**
- Give errors for common mutable usages of 'this' in readonly methods.
  - Field assignment
  - Non-readonly property setters
  - Passing by 'ref'
- Warn on new scenarios where an implicit copy is possible
  - Invoking a non-readonly instance member on 'this' from a readonly member

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 1]
```diff
@@ -328,8 +363,94 @@ static void Main()
...
+        P3 = 2; // ok
+        c.P = 42; // ok
+    }
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics(
+                // (27,9): error CS1604: Cannot assign to 'this' because it is read-only
+                //         P2 = 2; // error
+                Diagnostic(ErrorCode.ERR_AssgReadonlyLocal, "P2").WithArguments("this").WithLocation(27, 9));
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: This part concerns me a little bit. We could maybe communicate better that this property needs to be `readonly` in order to be assignable from a `readonly` method. But it requires either 1. passing more state into CheckValueKind or 2. giving errors at callsites instead of within CheckValueKind.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: I'm planning a future PR where there will be better diagnostics for several scenarios, particularly in places where I currently just give "readonly is not a valid modifier" (BadMemberFlag).

**Comment by RikkiGibson** [comment]
@dotnet/roslyn-compiler This PR is ready for review.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 2]
```diff
@@ -459,12 +459,10 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
                     // Note: RValueOnly is checked at the beginning of this method. Since we are here we need more than readable.
                     //"this" is readonly in members of readonly structs, unless we are in a constructor.
                     if (!thisref.Type.IsValueType ||
-                            (RequiresAssignableVariable(valueKind) &&
-                             thisref.Type.IsReadOnly &&
-                             (this.ContainingMemberOrLambda as MethodSymbol)?.MethodKind != MethodKind.Constructor))
+                            (RequiresAssignableVariable(valueKind) && (this.ContainingMemberOrLambda as MethodSymbol)?.IsEffectivelyReadOnly == true))
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: >         [](start = 20, length = 8)

Not part of this change but the indenting looks off.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 3]
```diff
@@ -459,12 +459,10 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
                     // Note: RValueOnly is checked at the beginning of this method. Since we are here we need more than readable.
                     //"this" is readonly in members of readonly structs, unless we are in a constructor.
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: >"this" is readonly in members of readonly structs, unless we are in a constructor [](start = 22, length = 81)

Does this comment need updating?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Yeah, I should update it to describe the the new readonly members scenario.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 4]
```diff
@@ -459,12 +459,10 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
...
                     if (!thisref.Type.IsValueType ||
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: >thisref.Type.IsValueType [](start = 25, length = 24)

Consider extracting a local. And perhaps remove the `thisref` local since it's only used to calculate `IsValueType`.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Done

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs (hunk 4 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Done

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 5]
```diff
@@ -289,6 +304,26 @@ public struct S
...
+        GetType();
+        ToString();
+        GetHashCode();
+        Equals(null);
+    }
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics();
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: >} [](start = 8, length = 1)

Please test with overridden `ToString()`, `GetHashCode()`, `Equals()`, with calls through `this` and `base`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 6]
```diff
@@ -817,6 +1025,32 @@ public struct S
...
+        s = default; // ok
+    }
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics(
+                // (12,9): error CS0191: A readonly field cannot be assigned to (except in a constructor or a variable initializer)
+                //         s.i = 42; // error
+                Diagnostic(ErrorCode.ERR_AssgReadonly, "s.i").WithLocation(12, 9));
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: >} [](start = 8, length = 1)

Please add tests for methods that are called implicitly:
- user-defined operators
- `foreach` and `GetEnumerator()`
- `using` and `Dispose()`

Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: >} [](start = 8, length = 1)

Can iterator and async methods be marked as `readonly`? Are we testing those cases?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: They're allowed in `readonly struct` so it seems like they should be allowed here. I'd like to test them. Is there anything significant to cover besides "this is still readonly in the method body"?

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs (hunk 5 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: Done

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs (hunk 6 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: - Agree there's not much more to test with user-defined operators because they're static methods.
- Will test that `foreach (var x in this)` warns appropriately within a readonly method
- Will test that `using (this) {}` warns appropriately within a readonly method
- Will test that deconstruction warns appropriately within a readonly method

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: May need to also test `await foreach` and `DisposeAsync`

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 7]
```diff
@@ -289,6 +304,93 @@ public struct S
             comp.VerifyDiagnostics();
         }
 
+        [Fact]
+        public void ReadOnlyMethod_CallBaseMethod()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: >ReadOnlyMethod_CallBaseMethod [](start = 20, length = 29)

Minor point: perhaps split into 4 separate tests so it's clear which warnings apply to each.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: Done

**Comment by RikkiGibson** [comment]
Remaining:

- [ ] check `readonly` iterator method
- [ ] check `readonly` async method
- [ ] check `readonly ref T GetPinnableReference()`

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs (hunk 6 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: Done

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: Done

**Comment by RikkiGibson** [comment]
Finished responding to feedback. Please review @dotnet/roslyn-compiler.

**Comment by 333fred** [comment]
>         var copy = this;

Do you have tests for ref members? ie, if I were to take a ref of `this`: `ref S s = ref this; s.M2()`

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadonlyStructTests.cs:1326 in e16f2e5. [](commit_id = e16f2e5c7b3e2682274e13056e1ebbf9c3920a1c, deletion_comment = False)

**Comment by RikkiGibson** [comment]
>         var copy = this;

See ReadOnlyStructsTests.ReadOnlyMethod_PassThisByRef

---
In reply to: [474116071](https://github.com/dotnet/roslyn/pull/34022#issuecomment-474116071) [](ancestors = 474116071)

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadonlyStructTests.cs:1326 in e16f2e5. [](commit_id = e16f2e5c7b3e2682274e13056e1ebbf9c3920a1c, deletion_comment = False)

**Comment by RikkiGibson** [comment]
>         var copy = this;

I'll add a test case for a scenario more like what you're describing though. Probably also a `ref readonly S s = ref this;` which we expect to compile.

---
In reply to: [474117236](https://github.com/dotnet/roslyn/pull/34022#issuecomment-474117236) [](ancestors = 474117236,474116071)

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadonlyStructTests.cs:1326 in e16f2e5. [](commit_id = e16f2e5c7b3e2682274e13056e1ebbf9c3920a1c, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 8]
```diff
@@ -289,6 +304,255 @@ public struct S
...
+    {
+        yield return i;
+        i++;
+        yield return i;
+    }
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics(
+                // (16,9): error CS1604: Cannot assign to 'this' because it is read-only
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: I really don't like this error message: I'm not trying to assign to `this`. I'm trying to assign to `i`, which is a member of `this`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 9]
```diff
@@ -289,6 +304,255 @@ public struct S
...
+                //         i++;
+                Diagnostic(ErrorCode.ERR_AssgReadonlyLocal, "i").WithArguments("this").WithLocation(16, 9));
+        }
+
+        [Fact]
+        public void ReadOnlyMethod_AsyncStreams()
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: Nit: consider putting this test with the `GetEnumerator` and family tests, as they're all tests of various `foreach` things.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 10]
```diff
@@ -289,6 +330,255 @@ public struct S
...
+{
+    static int i = 0;
+    readonly ref int GetPinnableReference() => ref i;
+    void M1()
+    {
+        fixed (int *i = this) {} // ok
+    }
+    readonly void M2()
+    {
+        fixed (int *i = this) {} // ok
```
Comment (agocke) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: The fact that this works at all is crazy

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs (hunk 8 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: You're right that this message should be better. Will mark it prototype so it can be addressed in a future PR.

Compare to when a field is marked `readonly`: https://sharplab.io/#v2:D4AQDABCCMDcCwAoEBmCAnApgQwCYHsA7AGwE8IBnAF3QFcBjKiAZSQG8kIuMcCTyAloSYCEiblAAsEALIAKAJSduHcRO4CA1JrESAvkj1A=

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 11]
```diff
@@ -328,8 +592,203 @@ static void Main()
...
+    {
+        get => 42;
+        set {}
+    }
+
+    public readonly void M()
+    {
+        P1 = 1; // ok
+        P2 = 2; // error
+        P3 = 2; // ok
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: This may have been LDM's decision, but this is my official objection to that decision. A `readonly setter` is an oxymoron.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 12]
```diff
@@ -817,6 +1305,211 @@ public struct S
...
         }
 
+        [Fact]
+        public void ReadOnlyStruct_Constructor()
+        {
+            var csharp = @"
+public readonly struct S
+{
+    public readonly int i;
+    public S(int i)
```
Comment (agocke) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: Is `readonly` prohibitied on constructors? Do we have a test for that?

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 13]
```diff
@@ -774,16 +1233,16 @@ public struct S
...
         {
-            // PROTOTYPE: would be good to add some more mutation here
-            // as well as expected diagnostics once that part of the feature is ready.
             var csharp = @"
 public struct S
 {
     private static int f1;
-    public readonly ref int M1() => ref f1;
+    public readonly ref int M1_1() => ref f1; // ok
+    public readonly ref readonly int M1_2() => ref f1; // ok
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: >readonly ref readonly [](start = 11, length = 21)

This is C++ `const`-level bad. Really disagreeing with LDM here.

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: And for the record: when we were talking about this feature offline, I got which `readonly` meant what wrong.

---
In reply to: [266668722](https://github.com/dotnet/roslyn/pull/34022#discussion_r266668722) [](ancestors = 266668722)

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs (hunk 12 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: See ReadOnlyStructsTests.ReadOnlyConstructor

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs [hunk 14]
```diff
@@ -1135,30 +1137,27 @@ private void CheckRestrictedTypeReceiver(BoundExpression expression, Compilation
...
-        private bool IsBindingModuleLevelAttribute()
+        /// <summary>
+        /// Returns false if an implicit 'this' copy will occur due to an instance member invocation in a readonly member.
+        /// </summary>
+        internal bool CheckImplicitThisCopyInReadOnlyMember(BoundExpression receiver, MethodSymbol method, DiagnosticBag diagnostics)
         {
-            if ((this.Flags & BinderFlags.InContextualAttributeBinder) == 0)
+            // For now we are warning only in implicit copy scenarios that are only possible with readonly members.
+            // Eventually we will warn on implicit value copies in more scenarios. See https://github.com/dotnet/roslyn/issues/33968.
+            if (receiver is BoundThisReference &&
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs]: This concerns me. What about something like this: `((InterfaceWithProp)this).Prop = 1;`

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs]: I'm curious whether boxing conversions on `this` within a readonly method should be a warning in general. Maybe only implicit ones?

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs (hunk 9 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: Done

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs [hunk 15]
```diff
@@ -1135,30 +1137,27 @@ private void CheckRestrictedTypeReceiver(BoundExpression expression, Compilation
...
+                receiver.Type.IsValueType &&
+                ContainingMemberOrLambda is MethodSymbol containingMethod &&
+                containingMethod.IsEffectivelyReadOnly &&
+                // Ignore calls to base members.
+                TypeSymbol.Equals(containingMethod.ContainingType, method.ContainingType, TypeCompareKind.ConsiderEverything) &&
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs]: Consider doing an assert that the containing method is `System.Object`. For now, the only thing that goes through this path is `struct`s which can only inherit from that, but if `readonly` members were to come to classes they'd likely break this assumption.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs]: Maybe that `containingMethod.ContainingType` is either equal to `method.ContainingType` or equal to `System.ValueType`?
